### PR TITLE
Elbv3 flavor param support

### LINKED ISSUE #73: Implement list Flavor handling

### LINKED ISSUE #76: Flavor validation error for valid flavor
Input
openstack --os-cloud octavia-nl loadbalancer create --name test-octavia-elb-ab2f551d_max --description "testing max parameters" --vip-subnet-id 54901817-e697-4666-b49d-50a74a2ad142 --availability-zone "eu-nl-02" --flavor 7628a037-c229-4c0c-820b-4ea862743aef --debug

Output
octaviaclient.api.exceptions.OctaviaClientException: Validation failure: Invalid flavor_id. (HTTP 400) (Request-ID: req-a41dd07c-306e-44c1-be8f-b49ceaad735b)
clean_up CreateLoadBalancer: Validation failure: Invalid flavor_id. (HTTP 400) (Request-ID: req-a41dd07c-306e-44c1-be8f-b49ceaad735b)
END return value: 1

### PR DESCRIPTION
**Pull request by anton-sidelnikov**
Closes #76
Closes #73

### CODE REVIEW

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 1]
```diff
@@ -113,10 +115,38 @@ def loadbalancer_create(self, session, loadbalancer):
...
         if 'tags' in lb_attrs:
             lb_attrs['tags'] = self._resource_tags(lb_attrs['tags'])
-        lb_attrs.pop('loadbalancer_id', None)
-        lb_attrs = elbv3_foremapping(lb_attrs)
+
+        if 'flavor_id' in lb_attrs:
+            flavors = []
+            if re.match(r's\d.\w+', lb_attrs['flavor_id']):
+                for item in ['L7', 'L4']:
+                    flavors.extend(session.vlb.find_flavor(
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:138]: ok, so here we have 2 calls to the API trying to get flavor with this name, fail, fetch list with filter resulting in overall 4 API calls. This is not really efficient

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 2]
```diff
@@ -113,10 +115,38 @@ def loadbalancer_create(self, session, loadbalancer):
...
+        lb_attrs['availability_zone_list'] = azs.replace(' ', '').split(',')
+
         if 'tags' in lb_attrs:
             lb_attrs['tags'] = self._resource_tags(lb_attrs['tags'])
-        lb_attrs.pop('loadbalancer_id', None)
-        lb_attrs = elbv3_foremapping(lb_attrs)
+
+        if 'flavor_id' in lb_attrs:
+            flavors = []
+            if re.match(r's\d.\w+', lb_attrs['flavor_id']):
```
[the remark this answers is not in the document]
Comment (gtema) [reply octavia_proxy/api/drivers/elbv3/driver.py:136]: We actually never going to get name as flavor_id - we only get ID or nothing therefore this branch should be useless

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 3]
```diff
@@ -113,10 +115,38 @@ def loadbalancer_create(self, session, loadbalancer):
...
+                for item in ['L7', 'L4']:
+                    flavors.extend(session.vlb.find_flavor(
+                        name_or_id=f'{item}_flavor.elb.{lb_attrs["flavor_id"]}'
+                    ))
+            else:
+                flavors.extend(session.vlb.get_flavor(lb_attrs['flavor_id']))
+                flavors.extend(session.vlb.find_flavor(
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:143]: This is way to complex to understand it without knowing the context. Please add very descriptive comment about our agreement on handling this.

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 4]
```diff
@@ -559,19 +589,31 @@ def flavors(self, session, project_id, query_filter=None):
...
 
         result = []
-
-        for fl in session.vlb.flavors(**query_filter):
-            fl_data = _flavors.FlavorResponse.from_sdk_object(fl)
-            fl_data.provider = PROVIDER
-            result.append(fl_data)
-
+        if 'name' in query_filter:
+            query_filter['name'] = ''.join(
```
[the remark this answers is not in the document]
Comment (gtema) [reply octavia_proxy/api/drivers/elbv3/driver.py:593]: please also add comment describing why you prepend L7_flavor.elb

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 5]
```diff
@@ -559,19 +589,31 @@ def flavors(self, session, project_id, query_filter=None):
...
+        if 'id' in query_filter:
+            fl_data = self.flavor_get(
+                project_id=project_id, session=session,
+                fl_id=query_filter['id']
+            )
+            if fl_data:
+                result.append(fl_data)
+        else:
+            for fl in session.vlb.flavors(**query_filter):
+                if not fl['name'].startswith('L4'):
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:605]: let's be friendly to anybody reading the code and use `if fl['name'].startswith('L4_flavor.elb'):` cause that will correspond to the replacement in the create_lb

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 6]
```diff
@@ -559,19 +589,31 @@ def flavors(self, session, project_id, query_filter=None):
...
+                    fl['name'] = re.sub(r'L\d_\w+.\w+.', '', fl['name'])
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:606]: then here you could just use:
```
  fl['name'][14:]
```

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 7]
```diff
@@ -559,19 +589,31 @@ def flavors(self, session, project_id, query_filter=None):
...
+                    result.append(fl_data)
         return result
 
-    def flavor_get(self, session, fl_id):
+    def flavor_get(self, session, project_id, fl_id):
         LOG.debug('Searching flavor')
-
         fl = session.vlb.find_flavor(
             name_or_id=fl_id, ignore_missing=True)
+        fl['name'] = re.sub(r'L\d_\w+.\w+.', '', fl['name'])
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:616]: don't know how this ended up here, but XXX_get is not for search by name, but only fetch using ID. This means no find_flavor and no regexps

File: octavia_proxy/api/v2/controllers/base.py [hunk 8]
```diff
@@ -93,7 +93,7 @@ def find_health_monitor(self, context, id, is_parallel=True):
 
         if not hm:
             raise exceptions.NotFound(
-                resource='Healthmonitor',
+                resource='HealthMonitor',
```
Comment (gtema) [inline octavia_proxy/api/v2/controllers/base.py:96]: does not seem related

File: octavia_proxy/api/v2/controllers/base.py [hunk 9]
```diff
@@ -117,11 +117,23 @@ def find_l7rule(self, context, l7policy_id, id, is_parallel=True):
 
         if not l7rule:
             raise exceptions.NotFound(
-                resource='L7 rule',
+                resource='L7Rule',
```
Comment (gtema) [inline octavia_proxy/api/v2/controllers/base.py:120]: this is also unrelated

File: octavia_proxy/tests/unit/api/drivers/elbv3/test_driver.py [hunk 10]
```diff
@@ -217,6 +259,11 @@ def test_flavors_no_qp(self):
         self.driver.flavors(self.sess, 'p1')
         self.sess.vlb.flavors.assert_called_with()
 
+    def test_flavor_get(self):
+        self.driver.flavor_get(self.sess, 'test', self.flavor)
+        self.sess.vlb.find_flavor.assert_called_with(
+            name_or_id=self.flavor, ignore_missing=True)
+
```
Comment (gtema) [inline octavia_proxy/tests/unit/api/drivers/elbv3/test_driver.py:267]: too much of new logic and too few tests. Would you please add following:
- list (drivers gives us returns multiple entries and we only return L7 ones)
- get (knowing concrete ID)
- find (passing `s1.medium`)
- find (passing `L7_flavor.elb.s1.medium`) to ensure we return nothing

File: octavia_proxy/api/drivers/elbv3/driver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py:136]: i added this for picking flavor by name:
openstack loadbalancer create --vip-subnet-id 34cbf900-68e1-47ce-854c-f3b98248bab9 --name lb-to-die --project c742c92afd8d46b1b3083d004afffd70 --flavor s2.large

[the remark this answers is not in the document]
Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py:136]: ok, will remove that, it didn't fail because validation happens bit later)

File: octavia_proxy/api/drivers/elbv3/driver.py (hunk 3 above)
Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py]: done

**Comment by gtema** [comment]
recheck
